aleemer/Internify
# [IN-102 and IN-111] Bug/in111 remove duplicate jobs

#### User Story
**IN-111 No Job duplicates in user's Job list**, and also **IN-102** (a fix/ change impacted both stories)
https://chaosneutral.atlassian.net/jira/software/projects/IN/boards/1/backlog?selectedIssue=IN-111

Also, https://chaosneutral.atlassian.net/jira/software/projects/IN/boards/1?selectedIssue=IN-102

Note @aleemer also contributed to this PR

#### Changes made
There are two categories of fixes implemented: 

(1) Display fix: implemented in Profile.js page, ensures that only unique jobs are displayed on the site.  In a sense it is an insurance policy; even if the bug is still at large (and there are duplicates in the store/ db), the user will not be aware of the bug as only unique job posts will be displayed in the UI.  

(2) Backend/ store fix:  UPDATE_JOBS_OF_USER has been modified within the userReducer to ensure that state.user.jobPostings is an array of unique members.  This is done through a filter function.  Viewing the Store (through redux middleware) I was able to verify that the filter function does work as desired, i.e. after it is invoked, by creating a new job, duplicates that once existed in the store are eliminated.

In addition, in the backend route of users (line 112), modification has been made such that only unique jobPostings are added, through replacing $push with $addToSet (See article: https://stackoverflow.com/questions/15921700/mongoose-unique-values-in-nested-array-of-objects).

The issue with these solutions is that though they address the effects of the duplicate bug problem I am unsure whether they prevent the bug from appearing in the first place.  This is because the bug only sporadically appears so it's difficult to assess without extensive automated testing over longer periods of time. 

Please note, there was a change made to user action's updateJobsOfUser; the payload was modified from **data.jobId** to just **data**.  Passing data.jobId was causing a "read property of undefined" warning in the console (See picture below). Using data instead eliminated this warning.

![photo_2021-08-08 18 00 32](https://user-images.githubusercontent.com/65432928/128649591-9b76daa8-d113-4087-9a4f-f04d29c787fc.jpeg)



#### Does your new code introduce new warnings on dev console?
No.

#### Test Steps
Create and edit jobs on your profile; you should not ever see any duplicate job posts
